monitor/events.go

```diff
@@ -3,76 +3,263 @@ package monitor
+	var sincePrevious time.Duration
+	now := time.Now()
+	if metrics.lastSegmentNonce == nonce {
+		sincePrevious = now.Sub(metrics.lastSegmentEmergedAt)
+	} else {
+		metrics.segmentsInFlight = 0
+		metrics.lastSegmentNonce = nonce
+		metrics.lastSeqNo = int64(seqNo) - 1
+	}
+	metrics.lastSegmentEmergedAt = now
+	e := &event{
+		Name:  "SegmentEmerged",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"seqNo":         seqNo,
+			"sincePrevious": uint64(sincePrevious / time.Millisecond),
+		},
+	}
+
+	sendPost(e)
+}
+
+func SegmentUploadStart(nonce, seqNo uint64) {
+	if metrics.lastSegmentNonce == nonce {
+		metrics.segmentsInFlight++
+	}
+}
+
+func LogSegmentUploaded(nonce, seqNo uint64, uploadDur time.Duration) {
+	e := &event{
+		Name:  "SegmentUploaded",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"seqNo":          seqNo,
+			"uploadDuration": uint64(uploadDur / time.Millisecond),
+		},
+	}
+
+	sendPost(e)
+}
+
+func detectSeqDif(e *event, nonce, seqNo uint64) {
+	if metrics.lastSegmentNonce == nonce {
+		seqDif := int64(seqNo) - metrics.lastSeqNo
+		metrics.lastSeqNo = int64(seqNo)
+		if seqDif != 1 {
+			e.Properties["seqNoDif"] = seqDif
+		}
+	}
+}
+
+func LogSegmentTranscoded(nonce, seqNo uint64, transcodeDur, totalDur time.Duration) {
+	glog.Infof("Logging LogSegmentTranscoded...")
+	if metrics.lastSegmentNonce == nonce {
+		metrics.segmentsInFlight--
+	}
+	e := &event{
+		Name:  "SegmentTranscoded",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"seqNo":             seqNo,
+			"transcodeDuration": uint64(transcodeDur / time.Millisecond),
+			"totalDuration":     uint64(totalDur / time.Millisecond),
+			"segmentsInFlight":  metrics.segmentsInFlight,
+		},
+	}
+	if metrics.segmentsInFlight != 0 {
+		e.Properties["segmentsInFlight"] = metrics.segmentsInFlight
+	}
+	detectSeqDif(e, nonce, seqNo)
+
+	sendPost(e)
+}
+
+func LogSegmentTranscodeFailed(nonce, seqNo uint64, reason string) {
+	glog.Infof("Logging LogSegmentTranscodeFailed...")
+	if metrics.lastSegmentNonce == nonce {
+		metrics.segmentsInFlight--
+	}
+	e := &event{
+		Name:  "SegmentTranscodeFailed",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"reason":           reason,
+			"seqNo":            seqNo,
+			"segmentsInFlight": metrics.segmentsInFlight,
+		},
+	}
+	detectSeqDif(e, nonce, seqNo)
+
+	sendPost(e)
+}
+
+func LogStartBroadcastClientFailed(nonce uint64, serviceURI, transcoderAddress string, jobID uint64, reason string) {
+	glog.Infof("Logging LogStartBroadcastClientFailed...")
+	e := &event{
+		Name:  "StartBroadcastClientFailed",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"jobID":             jobID,
+			"serviceURI":        serviceURI,
+			"reason":            reason,
+			"transcoderAddress": transcoderAddress,
+		},
+	}
+
+	sendPost(e)
+}
+
+func sendPost(e *event) {
```

Comment:
This could be refactored into `sendPost(name string, nonce uint64, props map[string]interface{})` which would save the event-creation boilerplate in each function